### PR TITLE
Overlays

### PR DESCRIPTION
**Pull request by hristoterezov**
WIP.
Done:
Refactors the overlays in Jitsi Meet to use react.
TODO:
Fix for filmstrip only mode.

### CODE REVIEW

File: modules/UI/UI.js [hunk 1]
```diff
@@ -40,6 +35,13 @@ import FollowMe from "../FollowMe";
 var eventEmitter = new EventEmitter();
 UI.eventEmitter = eventEmitter;
 
+/**
+ * Whether an overlay is visible or not.
+ * @type {boolean}
+ * @public
```
Comment (saghul) [inline modules/UI/UI.js]: does this really need to be public? looks like the `isOverlayVisible` getter should be used for checking and could be renamed to `_overlayVisible` and made private.

File: react/features/overlay/actionTypes.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import { Symbol } from '../base/react';
+
+/**
+ * The type of the Redux action which signals that a suspend was detected.
+ *
+ * {
+ *     type: SUSPEND_DETECTED
+ * }
+ * @public
```
Comment (saghul) [inline react/features/overlay/actionTypes.js]: I don't think we use visibility modifiers for action types.  So far they are always public.

File: react/features/overlay/actions.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+/**
+ * Signals that the prompt for media permission is visible or not.
+ *
+ * @param {boolean} isVisible - If the value is true - the prompt for media
+ * permission is visible otherwise the value is false/undefined.
+ * @param {string} browser - The name of the current browser.
+ * @returns {{
+ *     type: MEDIA_PERMISSION_PROMPT_VISIBILITY_CHANGED,
+ *     isVisible: {boolean},
+ *     browser: {string}
```
Comment (saghul) [inline react/features/overlay/actions.js]: is the browser info necessary here? the receiver of this action could query it using some utility function.

File: react/features/overlay/components/AbstractReloadableOverlay.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    constructor(props) {
+        super(props);
+
+        // Bind event handlers so they are only bound once for every instance.
+
+        /**
+         * @protected
+         * @type {Function}
+         */
+        this._reconnectNow = this._reconnectNow.bind(this);
```
Comment (saghul) [inline react/features/overlay/components/AbstractReloadableOverlay.js]: no need to do this since the function doesn't use `this`.

File: react/features/overlay/components/OverlayContainer.js [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+    /**
+     * Implements React's {@link Component#render()}.
+     *
+     * @inheritdoc
+     * @returns {ReactElement|null}
+     * @public
+     */
+    render() {
+        // FIXME: Temporary workaround until everything is moved to react.
+        APP.UI.overlayVisible
```
Comment (saghul) [inline react/features/overlay/components/OverlayContainer.js]: if we decide to go with a class component (see above) this can be done in lifecycle methods, which is kinda less hacky.

File: react/features/overlay/components/OverlayContainer.js [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+import { connect } from 'react-redux';
+import PageReloadOverlay from './PageReloadOverlay';
+import SuspendedOverlay from './SuspendedOverlay';
+import UserMediaPermissionsOverlay from './UserMediaPermissionsOverlay';
+
+/**
+ * Implements a React Component that will display the correct overlay when
+ * needed.
+ */
+class OverlayContainer extends Component {
```
Comment (saghul) [inline react/features/overlay/components/OverlayContainer.js]: At the moment his component only relies on props, there is no state nor lifecycle method usage so you can transform it into a stateless functional component.

File: react/features/overlay/components/PageReloadOverlay.js [hunk 7]
```diff
@@ -0,0 +1,292 @@
+/* global APP, AJS */
+import React, { Component } from 'react';
+import { randomInt } from '../../base/util/randomUtil';
+import AbstractReloadableOverlay from './AbstractReloadableOverlay';
+const logger = require('jitsi-meet-logger').getLogger(__filename);
```
Comment (saghul) [inline react/features/overlay/components/PageReloadOverlay.js]: does it not work with import?

File: react/features/overlay/components/PageReloadOverlay.js [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+     * @inheritdoc
+     * @returns {ReactElement|null}
+     * @public
+     */
+    render() {
+        return (
+            <div>
+                <div
+                    className = 'aui-progress-indicator'
+                    id = 'reloadProgressBar'>
```
Comment (saghul) [inline react/features/overlay/components/PageReloadOverlay.js]: Maybe the id would be an instance variable or a prop?

File: react/features/overlay/components/PageReloadOverlay.js [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+    /**
+     * Constructs overlay body with the warning message and count down towards
+     * the conference reload.
+     *
+     * @returns {ReactElement|null}
+     * @override
+     * @protected
+     */
+    _renderOverlayContent() {
+        const onFinish = this._reconnectNow;
```
Comment (saghul) [inline react/features/overlay/components/PageReloadOverlay.js]: why is this needed? can't you just inline it?

File: react/features/overlay/components/SuspendedOverlay.js [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    /**
+     * Constructs overlay body with the message and a button to rejoin.
+     *
+     * @returns {ReactElement|null}
+     * @override
+     * @protected
+     */
+    _renderOverlayContent() {
+        const btnClass = 'inlay__button button-control button-control_primary';
+        const onClickHandler = this._reconnectNow;
```
Comment (saghul) [inline react/features/overlay/components/SuspendedOverlay.js]: ditto

File: react/features/overlay/components/UserMediaPermissionsOverlay.js [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+     */
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            /**
+             * The src value of the image for the policy logo.
+             * @type {string}
+             */
+            policyLogoSrc: interfaceConfig.POLICY_LOGO
```
Comment (saghul) [inline react/features/overlay/components/UserMediaPermissionsOverlay.js]: doesn't look like this can ever change throughout the lifetime of the app, thus you can make it an instance variable.

File: react/features/overlay/reducer.js [hunk 12]
```diff
@@ -0,0 +1,142 @@
+/* global JitsiMeetJS */
+import {
```
Comment (saghul) [inline react/features/overlay/reducer.js]: order imports alphabetically

File: react/features/overlay/reducer.js [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+ * reduction of the specified action.
+ * @private
+ */
+function _conferenceFailed(state, action) {
+    const ConferenceErrors = JitsiMeetJS.errors.conference;
+
+    if (action.error === ConferenceErrors.FOCUS_LEFT
+        || action.error === ConferenceErrors.VIDEOBRIDGE_NOT_AVAILABLE) {
+        return setStateProperties(state, {
+            'haveToReload': true,
```
Comment (saghul) [inline react/features/overlay/reducer.js]: no need for quotes

File: react/features/overlay/reducer.js [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+ * Reduces a specific Redux action CONNECTION_ESTABLISHED of the feature
+ * overlay.
+ *
+ * @param {Object} state - The Redux state of the feature overlay.
+ * @returns {Object} The new state of the feature overlay after the
+ * reduction of the specified action.
+ * @private
+ */
+function _connectionEstablished(state) {
+    return setStateProperty(state, 'connectionEstablished', true);
```
Comment (saghul) [inline react/features/overlay/reducer.js]: ditto

File: react/features/overlay/components/AbstractOverlay.js [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            /**
+             * Indicates the css style of the overlay. if true - lighter  and
+             * darker otherwise.
+             * @type {boolean}
+             */
+            isLightOverlay: false
```
Comment (saghul) [inline react/features/overlay/components/AbstractOverlay.js]: this doesn't change throughout the lifetime of the application, you can make it an instance variable or a prop.

File: connection.js [hunk 16]
```diff
@@ -4,6 +4,9 @@ const logger = require("jitsi-meet-logger").getLogger(__filename);
 import AuthHandler from './modules/UI/authentication/AuthHandler';
 import jitsiLocalStorage from './modules/util/JitsiLocalStorage';
 
+import { connectionEstablished, connectionFailed }
```
Comment (saghul) [inline connection.js]: style: break these into individual lines

File: react/features/overlay/components/PageReloadOverlay.js (hunk 9 above)
Comment (hristoterezov) [inline react/features/overlay/components/PageReloadOverlay.js]: Do you mean `onFinish = { this._reconnectNow }`?

It seems that we have a eslint rule that the name of the handler should start with "on".

File: modules/UI/UI.js (hunk 1 above)
Comment (hristoterezov) [inline modules/UI/UI.js]: Currently we are setting this variable from the OverlayContainer.

File: react/features/overlay/actionTypes.js (hunk 2 above)
Comment (hristoterezov) [inline react/features/overlay/actionTypes.js]: I think we have a rule to add visibility modifiers on everything.

File: react/features/overlay/components/OverlayContainer.js (hunk 5 above)
Comment (hristoterezov) [inline react/features/overlay/components/OverlayContainer.js]: I can move it to `componentDidUpdate`. WDYT?

File: connection.js [hunk 17]
```diff
@@ -67,6 +70,23 @@ function connect(id, password, roomName) {
         connection.addEventListener(
             ConnectionEvents.CONNECTION_FAILED, handleConnectionFailed
         );
+        connection.addEventListener(
+            ConnectionEvents.CONNECTION_FAILED, connectionFailedHandler);
+
+        function connectionFailedHandler (error, errMsg) {
```
Comment (hristoterezov) [inline connection.js]: Yes they are very similar but in order to share the code I should add additional arguments and ifs and then I'll need to bind them. I think the code is going to become complex and ugly.

File: react/features/overlay/actions.js (hunk 3 above)
Comment (hristoterezov) [inline react/features/overlay/actions.js]: Good idea! 

The browser field is coming from lib-jitsi-meet currently. Also currently we don't have such utilities in jitsi-meet. I think we need more general strategy for checking the browser in jitsi-meet.

Isn't it better to fix this when we add such utilities in jitsi-meet?

File: react/features/overlay/components/PageReloadOverlay.js (hunk 7 above)
Comment (hristoterezov) [inline react/features/overlay/components/PageReloadOverlay.js]: It will work but it's going to look like this:
```
import {getLogger} from "jitsi-meet-logger";
const logger = getLogger(__filename);
```
Which is longer. 

Also currently this the way how the logger is used by the files from the react folder.

File: react/features/overlay/components/PageReloadOverlay.js (hunk 8 above)
Comment (hristoterezov) [inline react/features/overlay/components/PageReloadOverlay.js]: Why?

If we are going to use variable isn't it better to make it const since it won't change?

File: react/features/overlay/components/UserMediaPermissionsOverlay.js (hunk 11 above)
Comment (hristoterezov) [inline react/features/overlay/components/UserMediaPermissionsOverlay.js]: I think the future plans for config and interfaceConfig properties is to store them in the redux store and to access them from there. If we store them now as state properties the refactoring later is going to be easier.

File: react/features/overlay/components/AbstractOverlay.js (hunk 15 above)
Comment (hristoterezov) [inline react/features/overlay/components/AbstractOverlay.js]: I think it's good to keep it in the state because this is abstract class and in future some of the classes that are going to extend it may need to change during their lifetime.

File: react/features/overlay/reducer.js (hunk 14 above)
Comment (hristoterezov) [inline react/features/overlay/reducer.js]: Are you sure? This is the `setStateProperty` method.

File: react/features/overlay/components/PageReloadOverlay.js (hunk 9 above)
Comment (hristoterezov) [inline react/features/overlay/components/PageReloadOverlay.js]: I'll disable the rule.

File: react/features/overlay/components/AbstractReloadableOverlay.js [hunk 18]
```diff
@@ -0,0 +1,20 @@
+/* global APP */
+import AbstractOverlay from './AbstractOverlay';
+
+/**
+ * Implements an abstract React Component for overlay with ability to
+ * reload the page.
+ */
+export default class AbstractReloadableOverlay extends AbstractOverlay {
```
Comment (saghul) [inline react/features/overlay/components/AbstractReloadableOverlay.js]: I'm not sure it's worth it to make an abstract component here, since all it's shared is a single one-line function which will need to be refactored anyway.  I'd remove this.

File: react/features/overlay/components/OverlayContainer.js (hunk 6 above)
Comment (saghul) [inline react/features/overlay/components/OverlayContainer.js]: My previous comment no longer applies since you now use `componentDidUpdate`.

File: react/features/overlay/components/OverlayContainer.js [hunk 19]
```diff
@@ -0,0 +1,212 @@
...
+    /**
+     * OverlayContainer component's property types.
+     *
+     * @static
+     */
+    static propTypes = {
```
Comment (saghul) [inline react/features/overlay/components/OverlayContainer.js]: There is something I don't like about these props.  They don't apply to the entire component, just to some of the sub-components.

Here is an alternative proposal: just create 2 props in mapStateToProps`: `overlayType` and `overlayOptions`.  Or, since we are taling about overlays, `type` and `options`.  Then in the render method you can create the right component based on the `type` prop and blindly pass the props in `options` using the spread operator. I think this would simplify the code quite a bit.

File: modules/UI/UI.js (hunk 1 above)
Comment (hristoterezov) [inline modules/UI/UI.js]: I think I'm missing something. How it can be private since it is used by another module? Do you want me to create a public setter for it? 

Or do you want to add "_" to the variable name just to mark this variable as something that shouldn't be used? If this is the case I prefer to fix this with more comments.

Comment (saghul) [inline modules/UI/UI.js]: > Or do you want to add "_" to the variable name just to mark this variable as something that shouldn't be used?

This.

File: react/features/overlay/actions.js (hunk 3 above)
Comment (hristoterezov) [inline react/features/overlay/actions.js]: Now or later we'll always need 3 changes. 

I would do it but there are some decisions to be made like:
1) should we export RTCBrowserType from lib-jitsi-meet and start using it from jitsi-meet
2) should we create js util package where we can put RTCBrowserType and use it in lib-jitsi-meet and jitsi-meet.
3) or maybe we should use 3rd party library for browser detection in jitsi-meet
Probably there are a lot other options.

Recently there was discussion about this topic but I don't think we have final decision yet. 

I don't see a point to make this change now if we have to fix it again later. Also the scope of the PR is to refactor the overlays to work with react and that's how they were working before. So I really prefer to fix this later.

File: react/features/overlay/components/AbstractOverlay.js (hunk 15 above)
Comment (hristoterezov) [inline react/features/overlay/components/AbstractOverlay.js]: Hmmm... Good point. Probably not very likely to be changed.

File: react/features/base/connection/actions.web.js [hunk 20]
```diff
@@ -8,6 +8,11 @@ import './reducer';
 const JitsiConferenceEvents = JitsiMeetJS.events.conference;
 const logger = require('jitsi-meet-logger').getLogger(__filename);
 
+export {
+    connectionEstablished,
```
Comment (hristoterezov) [inline react/features/base/connection/actions.web.js]: @lyubomir Do we have some convention about this? I'm OK with @saghul 's proposal but I remembered that generally the other team members were concerned about the number of files that we are creating/using.

File: react/features/overlay/components/AbstractReloadableOverlay.js (hunk 18 above)
Comment (hristoterezov) [inline react/features/overlay/components/AbstractReloadableOverlay.js]: The only reason for this was the rule in our code style guide "No duplicated code".

File: react/features/overlay/actionTypes.js (hunk 2 above)
Comment (lyubomir) [inline react/features/overlay/actionTypes.js]: After the introduction of `_` as a access modifier hint, we left all actionTypes without `_` which means that they're public. For the sake of consistency, people may feel that `@public` is necessary so I'm OK with it. It should be noted though that `@public` is missing pretty much everywhere and not only on actionTypes.

File: react/features/overlay/components/OverlayContainer.js (hunk 19 above)
Comment (hristoterezov) [inline react/features/overlay/components/OverlayContainer.js]: Where do you think the decision about the value of `overlayType` should be made? The reducer maybe?

File: modules/UI/UI.js (hunk 1 above)
Comment (hristoterezov) [inline modules/UI/UI.js]: I think this would be confusing because technically this variable is public. What about adding more comments?

File: react/features/base/connection/actions.web.js (hunk 20 above)
Comment (lyubomir) [inline react/features/base/connection/actions.web.js]: I don't like actions of a base, non-UI feature should be split between .web and .native. Temporarily though while we move more of the logic of the current Web into React, Illia chose to split them. I'll try very hard to remove the split here (and in functions.js where we have them currently).

File: react/features/overlay/actions.js (hunk 3 above)
Comment (saghul) [inline react/features/overlay/actions.js]: Ok.  FWIW I have used [bowser](https://www.npmjs.com/package/bowser) in the past and is quite nice,

File: react/features/overlay/components/OverlayContainer.js (hunk 19 above)
Comment (saghul) [inline react/features/overlay/components/OverlayContainer.js]: I was thinking about doing it in the mapping function, since it has all the knowledge, the reducer should work too.

File: modules/UI/UI.js (hunk 1 above)
Comment (saghul) [inline modules/UI/UI.js]: Ok.

File: react/features/overlay/components/AbstractReloadableOverlay.js (hunk 18 above)
Comment (saghul) [inline react/features/overlay/components/AbstractReloadableOverlay.js]: While that is true, it forces us to add an unnecessary abstraction for the purpose of charing one line of code...

File: conference.js [hunk 21]
```diff
@@ -20,6 +20,12 @@ import analytics from './modules/analytics/analytics';
 
 import EventEmitter from "events";
 
+import { conferenceFailed } from './react/features/base/conference';
+import {
+    suspendDetected,
+    mediaPermissionPromptVisibilityChanged
```
Comment (saghul) [inline conference.js]: style: sort alphabetically

File: modules/UI/UI.js (hunk 1 above)
Comment (saghul) [inline modules/UI/UI.js]: Ok then.  This is temporary anyway, right?  Then let's also mention that.

File: modules/UI/UI.js [hunk 22]
```diff
@@ -1403,10 +1391,7 @@ UI.hideRingOverLay = function () {
  * @returns {*|boolean} {true} if the overlay is visible, {false} otherwise
  */
 UI.isOverlayVisible = function () {
-    return RingOverlay.isVisible()
-        || SuspendedOverlay.isVisible()
-        || PageReloadOverlay.isVisible()
-        || GumPermissionsOverlay.isVisible();
+    return RingOverlay.isVisible() || this.overlayVisible;
```
Comment (saghul) [inline modules/UI/UI.js]: Question: what's so special about the ring overlay?  Should its migration be part of this PR?

File: react/features/app/components/AbstractApp.js [hunk 23]
```diff
@@ -1,3 +1,4 @@
+/* global APP */
```
Comment (saghul) [inline react/features/app/components/AbstractApp.js]: do we usually leave an empty line after this comment?

File: react/features/base/connection/actions.native.js [hunk 24]
```diff
@@ -180,13 +183,13 @@ function _connectionDisconnected(connection, message) {
  *
  * @param {JitsiConnection} connection - The JitsiConnection which was
  * established.
- * @private
  * @returns {{
  *     type: CONNECTION_ESTABLISHED,
  *     connection: JitsiConnection
  * }}
+ * @public
```
Comment (saghul) [inline react/features/base/connection/actions.native.js]: style: I think we tend to put this before `@returns`.

File: react/features/overlay/components/AbstractReloadableOverlay.js (hunk 18 above)
Comment (hristoterezov) [inline react/features/overlay/components/AbstractReloadableOverlay.js]: Actually I'll move the method to `AbstractOverlay` class.

I started implementing the filmstrip only version and it turned out that I have to create `AbstractReloadableOverlay` and also `AbstractOverlay` for filmstrip only mode also. That way  we have to have +2 more files only because of that method :) So you were right :)

File: modules/UI/UI.js (hunk 22 above)
Comment (hristoterezov) [inline modules/UI/UI.js]: Eventually it should be. The thing is that it was not extending Overlay.js and the refactoring will be a little bit harder that the other overlays.

File: react/features/base/connection/actions.native.js (hunk 24 above)
Comment (hristoterezov) [inline react/features/base/connection/actions.native.js]: I don't think we have rule for this. I personally think that params and returns should be together (I consider them like one group) and we can add the access modifier either after or before them.

Comment (lyubomir) [inline react/features/base/connection/actions.native.js]: I hope we don't postpone a decision on this long enough to end up with 2 different coding styles merely out of personal preference.

Comment (emcho) [inline react/features/base/connection/actions.native.js]: @lyubomir this is react so the decision is yours. @hristoterezov could you please comply? 
Thanks

File: connection.js [hunk 25]
```diff
@@ -67,6 +72,23 @@ function connect(id, password, roomName) {
...
+            APP.store.dispatch(connectionFailed(connection, error, errMsg));
+
+            switch (error) {
+            case ConnectionErrors.CONNECTION_DROPPED_ERROR:
+            case ConnectionErrors.OTHER_ERROR:
+            case ConnectionErrors.SERVER_ERROR: {
```
Comment (lyubomir) [inline connection.js]: Saúl commented on the fact that the 2 `CONNECTION_FAILED` handlers look almost duplicate and I understand how it may seem complex to reduce duplication. However, we have to agree that we have a `switch (error)` with the same group of 3 `case`s in 3 files just in this PR. Obviously, there's something special about these 3 `ConnectionErrors`. What's not obvious (to me) is what that special thing is. So it would've been both a duplication reduction and a clarification if the testing for these 3 `cases` was in a properly named function. It could've even been a method on `ConnectionErrors`.

File: react/features/overlay/components/PageReloadOverlay.js [hunk 26]
```diff
@@ -0,0 +1,298 @@
...
+
+const logger = require('jitsi-meet-logger').getLogger(__filename);
+
+/**
+ * Implements a React Component for the reload timer. Starts counter from
+ * props.start, adds props.step to the current value on every props.interval
+ * seconds until the current value reaches props.end. Also displays progress
+ * bar.
+ */
+class ReloadTimer extends Component {
```
Comment (lyubomir) [inline react/features/overlay/components/PageReloadOverlay.js]: I think it aids readability when `class`es are defined in their own files. `ReloadTimer` was extracted in its own `class` because it's a standalone abstraction anyway, right?

File: react/features/overlay/components/PageReloadOverlay.js [hunk 27]
```diff
@@ -0,0 +1,298 @@
+/* global APP, AJS */
+
+import React, { Component } from 'react';
+
+import { randomInt } from '../../base/util/randomUtil';
```
Comment (lyubomir) [inline react/features/overlay/components/PageReloadOverlay.js]: The exporting/importing of the `./react` features is package-based rather than file-based.

File: react/features/overlay/components/PageReloadOverlay.js [hunk 28]
```diff
@@ -0,0 +1,298 @@
...
+    /**
+     * React Component method that executes once component is mounted.
+     *
+     * @inheritdoc
+     * @returns {void}
+     * @protected
+     */
+    componentDidMount() {
+        AJS.progressBars.update('#reloadProgressBar', 0);
+        const intervalId = setInterval(() => {
```
Comment (lyubomir) [inline react/features/overlay/components/PageReloadOverlay.js]: You're starting a timer when the Component is mounted and letting it continue even after the Component is unmounted. While in the particular use of `ReloadTimer` that may not be a problem, it's safer and cleaner in general to clear the timer when the Component unmounts.

File: react/features/overlay/components/PageReloadOverlay.js [hunk 29]
```diff
@@ -0,0 +1,298 @@
...
+     *
+     * @param {Object} props - The read-only properties with which the new
+     * instance is to be initialized.
+     * @public
+     */
+    constructor(props) {
+        super(props);
+        this.state = {
+            current: this.props.start,
+            time: Math.abs(this.props.end - this.props.start)
```
Comment (lyubomir) [inline react/features/overlay/components/PageReloadOverlay.js]: These are pretty much like fields of the class and, consequently, it's nice to have jsdocs on them.

File: react/features/overlay/components/PageReloadOverlay.js [hunk 30]
```diff
@@ -0,0 +1,298 @@
...
+        super.componentDidMount();
+
+        // FIXME (CallStats - issue) this event will not make it to
+        // the CallStats, because the log queue is not flushed, before
+        // "fabric terminated" is sent to the backed
+        // FIXME: We should dispatch action for this
+        APP.conference.logEvent('page.reload', undefined /* value */,
+            this.props.reason /* label */);
+        logger.info(`The conference will be reloaded after
+            ${this.state.timeoutSeconds} seconds.`);
```
Comment (lyubomir) [inline react/features/overlay/components/PageReloadOverlay.js]: The template literal above does take the newline and all the spaces from the beginning of the second line into the resulting string. Which is not a real problem in this case, one just ends up with a funny looking info log line.

File: react/features/overlay/components/SuspendedOverlay.js [hunk 31]
```diff
@@ -0,0 +1,37 @@
...
+     * Constructs overlay body with the message and a button to rejoin.
+     *
+     * @returns {ReactElement|null}
+     * @override
+     * @protected
+     */
+    _renderOverlayContent() {
+        const btnClass = 'inlay__button button-control button-control_primary';
+
+        /* eslint-disable react/jsx-handler-names */
```
Comment (lyubomir) [inline react/features/overlay/components/SuspendedOverlay.js]: Do you know that `eslint-disable` will seize to be in effect as you expect? I'd rather we be safe rather than sorry and `eslint-enable` after a `eslint-disable`.

File: react/features/overlay/reducer.js [hunk 32]
```diff
@@ -0,0 +1,140 @@
...
+ *
+ * @param {Object} state - The Redux state of the feature overlay.
+ * @param {Action} action - The Redux action to reduce.
+ * @returns {Object} The new state of the feature overlay after the reduction of
+ * the specified action.
+ * @private
+ */
+function _mediaPermissionPromptVisibilityChanged(state, action) {
+    return setStateProperties(state, {
+        mediaPermissionPromptVisible: action.isVisible,
```
Comment (lyubomir) [inline react/features/overlay/reducer.js]: I'm afraid I don't follow the `boolean` variable naming style followed in this PR. Regardless of the side on which I stand, do you or do you not want to have `is`? I realize that `is` encodes the `boolean` type into the variable name but you're not using it always in your PR so I'm not sure whether you're taking advantage of the name, whether the rule is more complex than simply `typeof isXXX === 'boolean'`, or there's merely no rule.

File: react/features/overlay/components/ReloadTimer.js [hunk 33]
```diff
@@ -0,0 +1,141 @@
...
+        return (
+            <div>
+                <div
+                    className = 'aui-progress-indicator'
+                    id = 'reloadProgressBar'>
+                    <span className = 'aui-progress-indicator-value' />
+                </div>
+                <span
+                    className = 'reload_overlay_text'
+                    id = 'reloadSeconds'>
```
Comment (lyubomir) [inline react/features/overlay/components/ReloadTimer.js]: What is the purpose of `id = 'reloadSeconds'`? I searched for `reloadSeconds` and I only found it here.

File: react/features/overlay/components/SuspendedOverlay.js [hunk 34]
```diff
@@ -0,0 +1,39 @@
...
+            <div className = 'inlay'>
+                <span className = 'inlay__icon icon-microphone' />
+                <span className = 'inlay__icon icon-camera' />
+                <h3
+                    className = 'inlay__title'
+                    data-i18n = 'suspendedoverlay.title' />
+                <button
+                    className = { btnClass }
+                    data-i18n = 'suspendedoverlay.rejoinKeyTitle'
+                    id = 'rejoin'
```
Comment (lyubomir) [inline react/features/overlay/components/SuspendedOverlay.js]: What is the purpose of `id = 'rejoin'`? I searched for `rejoin` and I only found it here.

File: react/features/overlay/components/AbstractOverlay.js [hunk 35]
```diff
@@ -0,0 +1,91 @@
...
+        );
+    }
+
+    /**
+     * Reloads the page.
+     *
+     * @returns {void}
+     * @protected
+     */
+    _reconnectNow() {
```
Comment (lyubomir) [inline react/features/overlay/components/AbstractOverlay.js]: `_reconnectNow` is used by (at least 2) extenders as an event handler in the form `onEvent = { this._reconnectNow }`. Which means that they will not define a `this` inside `_reconnectNow`. Given that (1) such problems can be a bit surprising at first to figure out and (2) there's future work planned on the method, it may be easier on maintenance to `.bind(this)` in the constructor methods which are defined with the intention to be used as event handlers.

File: react/features/base/connection/actions.native.js (hunk 24 above)
Comment (hristoterezov) [inline react/features/base/connection/actions.native.js]: @emcho Sure! I was just expressing my personal preference for the new rule. But I don't agree that this is related to React especially if at some point we try to create a common JS code style guide for all our JS projects.

File: react/features/overlay/components/PageReloadOverlay.js (hunk 26 above)
Comment (hristoterezov) [inline react/features/overlay/components/PageReloadOverlay.js]: I was trying to reduce the number of files because `ReloadTimer` is used only here. But I accept your arguments. I don't mind if we put it in a new file.

File: react/features/overlay/components/PageReloadOverlay.js (hunk 28 above)
Comment (hristoterezov) [inline react/features/overlay/components/PageReloadOverlay.js]: Good catch! I've completely forgot about stopping it.

File: react/features/overlay/components/SuspendedOverlay.js (hunk 31 above)
Comment (hristoterezov) [inline react/features/overlay/components/SuspendedOverlay.js]: I don't know and I haven't thought about it.

I agree with you.

File: react/features/overlay/reducer.js (hunk 32 above)
Comment (hristoterezov) [inline react/features/overlay/reducer.js]: I don't have any preferences about a rule for naming boolean variables and in the PR simply there is no rule about it. 

Do we have a rule for this? Sorry if there was a rule that I broke.

File: react/features/overlay/components/ReloadTimer.js (hunk 33 above)
Comment (hristoterezov) [inline react/features/overlay/components/ReloadTimer.js]: Probably I was using it at some point and later forgot to remove it. I think it's safe to be removed now.

File: react/features/overlay/components/SuspendedOverlay.js (hunk 34 above)
Comment (hristoterezov) [inline react/features/overlay/components/SuspendedOverlay.js]: Oh that's leftovers from the previous version of the code that was not using React. Now it is really not used. We can remove it.

File: react/features/overlay/components/AbstractOverlay.js (hunk 35 above)
Comment (hristoterezov) [inline react/features/overlay/components/AbstractOverlay.js]: Initially I did `.bind(this)` for similar reasons. But it seems that @saghul prefers to not have `.bind(this)` because we are actually not using `this` there. 

I don't have any opinion on this matter and agree with the arguments of both of you.
